testing/unittests/datamaskingtests.cpp

```diff
@@ -0,0 +1,1649 @@
+    See the License for the specific language governing permissions and
+    limitations under the License.
+############################################################################## */
+
+#ifdef _USE_CPPUNIT
+#include "datamaskingplugin.hpp"
+#include "datamaskingengine.hpp"
+#include "datamaskingcompatibility.hpp"
+#include "unittests.hpp"
+#include <vector>
+
+struct SubstituteException
+{
+};
+struct ExpectedException : public SubstituteException
+{
+};
+
+ITraceProducer* createTracer()
+{
+  class Filter : public CInterfaceOf<IModularTraceMsgFilter>
+  {
+  public:
+      virtual bool rejects(const LogMsgCategory& category) const override
+      {
+        return ((category.queryClass() & classMask) != category.queryClass());
+      }
+    private:
+      int classMask = MSGCLS_error;
+    public:
+      Filter()
+      {
+      }
+      Filter(int _classMask)
+        : classMask(_classMask)
+      {
+      }
+  };
+  Owned<CModularTraceProducer> tracer(new CModularTraceProducer());
+  tracer->setSink(new CConsoleTraceMsgSink(true, true));
+  tracer->setFilter(new Filter());
+  //tracer->setFilter(new Filter(MSGCLS_error | MSGCLS_information));
+  return tracer.getLink();
+}
+
+class DataMaskingEngineTests : public CppUnit::TestFixture
+{
+    CPPUNIT_TEST_SUITE( DataMaskingEngineTests );
+        CPPUNIT_TEST(testMinimalPlugin);
+        CPPUNIT_TEST(testLegacyDomains);
+        CPPUNIT_TEST(testMultipleDomains);
+        CPPUNIT_TEST(testCheckCompatibility);
+    CPPUNIT_TEST_SUITE_END();
+
+    using Engine = DataMasking::CEngine;
+    using Checker = DataMasking::CCompatibilityChecker;
+
+    struct CurrentTest
+    {
+      DataMaskingEngineTests& tester;
+      CurrentTest(DataMaskingEngineTests& _tester, const char* test)
+          : tester(_tester)
+      {
+        tester.currentTest = test;
+      }
+      ~CurrentTest()
+      {
+        tester.currentTest = nullptr;
+      }
+    };
+    const char* currentTest = nullptr;
+
+public:
+    void testMinimalPlugin()
+    {
+      static const char* cfgTxt = R"!!!(
+        MaskingPlugin:
+          - library: datamasker
+            entry-point: newPartialMaskSerialToken
+            domain: 'urn:hpcc:unittest'
+      )!!!";
+
+      CurrentTest ct(*this, "testMinimalPlugin");
+      Owned<Engine> engine(createEngine(cfgTxt));
+
+      assertDomainCount(engine, 1);
+      assertHasDomain(engine, "urn:hpcc:unittest", true);
+      assertHasDomain(engine, "bad", false);
+      assertGetDomains(engine, {"urn:hpcc:unittest" });
+      assertGetDomainIds(engine, "urn:hpcc:unittest", { "urn:hpcc:unittest" });
+      assertDomainVersions(engine, "urn:hpcc:unittest", { 1 });
+      assertQueryProfile(engine, "", 0, true);
+    }
+
+    void testLegacyDomains()
+    {
+      static const char* cfgTxt = R"!!!(
+        MaskingPlugin:
+          library: datamasker
+          entry-point: newPartialMaskSerialToken
+          Profile:
```

Comment:
Quite trivial, but I think we should be standardizing our yaml style.  Take a look at some of the hpcc YAML files for indenting, case, etc.  Case can be tricky if you are trying to support both xml and yaml.  What is the primary format?

Reply:
Can you identify a file that illustrates what you have in mind? I see combinations of title case, camel case and snake case, and found a file that couldn't decide whether to indent one, two, or three spaces. I may be unique with hyphenated names, but I can't tell how to standardize it.

Reply:
The format can be either yaml or xml. I would expect yaml to be preferred if standalone, but accept XML so a single definition can either be deployed as a standalone file or included in the ESDL binding.

Reply:
The prime example especially from a user's configuration perspective is https://github.com/hpcc-systems/HPCC-Platform/blob/master/helm/hpcc/values.yaml.
Notice indent is 2 spaces.
The dash of an array entry lines up with the parent tag.
Names start with lower case and are camel case after that where appropriate.
Names with dashes could be up for discussion I wasn't as concerned about that but we should discuss it.